erigon-lib/commitment/hex_patricia_hashed.go

```diff
@@ -1566,6 +1566,37 @@ func (hph *HexPatriciaHashed) createCellGetter(b []byte, updateKey []byte, row,
 	}
 }
 
+// updateKind is a type of update that is being applied to the trie structure.
+type updateKind uint8
+
+const (
+	// updateKindDelete means after we processed longest common prefix, row ended up empty.
+	updateKindDelete updateKind = 0b0
+
+	// updateKindPropagate is an update operation ended up with a single nibble which is leaf or extension node.
+	// We do not store keys with only one cell as a value in db, instead we copy them upwards to the parent branch.
+	// IF updated prefix was present before, operation produces a delete write for it.
```

Comment:
Not sure what this means. Do you mean that in the grid if there is an extra row with only one cell, that cell is actually saved in the db as part of the row above it? (i.e. the row with the singleton is only in the grid presentation layer in memory but not in the db representation?)

Reply:
Also you make a reference to a delete operation. Isn't that a different kind of update?

Reply:
which is represented as updateKindDelete

Reply:
> Not sure what this means. Do you mean that in the grid if there is an extra row with only one cell, that cell is actually saved in the db as part of the row above it? (i.e. the row with the singleton is only in the grid presentation layer in memory but not in the db representation?)

yes as we previously talked, no singular cell is stored in db alone.

Reply:
>  which is represented as updateKindDelete

> IF updated prefix was present before, operation produces a delete write for it.

Could you explain this scenario? If the prefix was present before and there is an update to it, how does that end up being a delete operation?

Reply:
once branch was created with eg 2 cells. Then eg those 2 storage cells were deleted, branch become empty. Now to keep that fact in db we have to make tx.Delete(prefix, []byte{}) which is deletion write operation.

Reply:
update in general is a change to something. Therefore change to a branch ending up with 0 cells in it considered as "deletion update"

Reply:
Ok, then better to attach this explanation to the corresponding update type which is `UpdateKindDelete`.

The sentence : `// IF updated prefix was present before, operation produces a delete write for it.` sounds a bit misleading because if I understood correctly `updateKindPropagate` does not perform any deletion.

Reply:
`updateKindPropagate` means that we had prefix `0xdead` for leaf, now leaf moves up to `0xdea` branch prefix so `0xdead` must be removed.